Add checking of Markdown documents to our linters

krcroft opened the pull request:
This PR uses mdl (markdownlint) which has become a popular checker for Markdown files.

 - https://github.com/markdownlint/markdownlint (Ruby-based command-line version we're using)
 - https://github.com/DavidAnson/markdownlint (Node.JS HTTP-service based version)
 - https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint (and as a built-in checker for Visual Studio Code)

![2020-01-05_22-40](https://user-images.githubusercontent.com/1557255/71800309-9a054980-300c-11ea-87a3-8f54af38b7fb.png)

Links to our three current Markdown docs:
 - https://github.com/dreamer/dosbox-staging/blob/kc/verify-md-1/README.md
 - https://github.com/dreamer/dosbox-staging/blob/kc/verify-md-1/contrib/README.md
 - https://github.com/dreamer/dosbox-staging/blob/kc/verify-md-1/scripts/build.md

krcroft commented on .github/workflows/analysis.yml:
```diff
@@ -13,12 +13,16 @@ jobs:
+          sudo apt-get install python3-setuptools ruby
           sudo pip3 install pylint beautifulsoup4 html5lib
+          sudo gem install mdl
+          sudo chmod go-w /usr/share/rust/.cargo/bin
```
On startup, Ruby complain about a permission issue here - so this cleans that up in our output.

krcroft commented on .mdlrc:
```diff
@@ -0,0 +1 @@
+rules "~MD005", "~MD033", "~MD046"
```
33 will complain about embedded HTML and didn't like `<sup>`, so that had to go.
The other two were too picky and hard to satisfy.

dreamer commented on README.md:
```diff
@@ -45,18 +45,20 @@ Codecs supported for CD-DA emulation:
+
+<sup>‡ - 44.1 kHz stereo only</sup>
+
+<sup>§ - Broken or unsupported in either SDL\_sound or DOSBox</sup>
```
Previously these lines used 2 spaces to break the line, which I presumed was correct markdown… Does it generate any warning? In this case, we don't want every annotation to be in a separate paragraph.

krcroft replied:
Yes, separate paragraphs is unnecessary visual bloat.  Two (or more) spaces still generates:
- README.md:48: MD009 Trailing spaces
- README.md:49: MD009 Trailing spaces
- README.md:50: MD009 Trailing spaces

Edit: Interesting.. Markdown uses the more explicit "\" to create a line-break without starting a new paragraph, similar to <br>.  Will try that.

krcroft replied:
Edit 2: \\ works!

Makes me think we can now safely add a "strip-trailing-whitespace" commit-hook, because nothing I know of should have trailing whitespace.. can you think of anything?

dreamer replied:
Nope, breaking space was the only thing - and the new way is better than having 2 trailing spaces :)

dreamer commented on README.md:
```diff
@@ -68,18 +70,17 @@ right), and download the package.
-#### Fedora
-
-    $ sudo dnf install libpng SDL2 SDL2_net opusfile
+### Fedora
```
I think I made an error earlier - sections Linux, Windows, macOS should be subsections of "Development snapshot builds", and this would keep sections per distro at level 4, otherwise it might be visually hard to distinguish where next section starts.

Change it to:
```
### [Linux](…)

…

#### Fedora

…

### [Windows](…)

…
```
This way linter probably won't complain about nesting structure.

krcroft replied:
Will do!

krcroft replied:
Yes, nice and consistent now.

dreamer commented on README.md:
```diff
@@ -107,11 +107,11 @@ Read [INSTALL](INSTALL) file for a general summary about dependencies and
+    cd dosbox-staging
+    ./autogen.sh
+    ./configure
+    make
```
Maybe shell syntax highlight? :)

krcroft replied:
Will do - but I do like how tidy the raw text looked in your original version :-)
The linter didn't like the $ prefixes (justified by the them making it harder to copy and paste).

krcroft replied:
Fenced, along with a couple others.

dreamer commented on scripts/verify-markdown.sh:
```diff
@@ -0,0 +1,29 @@
+# SPDX-License-Identifier: GPL-2.0-or-later
+#
+# Based on `verify-bash.sh` by Patryk Obara <patryk.obara@gmail.com>
+# Copyright (C) 2019 licensed GPL-2.0-or-later
```
Eh, you can remove me - additional copyright line won't cause problems for scripts updating years in the future :).

krcroft replied:
sounds good; yeah, was just trying to show attribution - but yeah, too much of that would get cluttered!

krcroft replied:
Done.

krcroft commented:
Squashed and pushed.
Turns out powershell can be highlighted too (https://github.com/MicrosoftDocs/PowerShell-Docs/issues/1511), so I fenced that.  Changed to `PS>` to `PS:\>` because that issue mentions GitHub will skip that prefix when users copy and paste from a `powershell` fenced block. (pretty cool).
